Export new episodes

Smeulf opened the pull request:
"Breacking" change: TvShows or Seasons exported with NFO will need a refresh for the export new episodes feature to export NFOs, or by default NFO won't be exported

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the [**CONTRIBUTING**](Contributing.md) document.

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?


### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [X] Have you successfully ran tests with your changes locally?

Smeulf commented:
It's not squashed for better readability, but needs to be at the end.

CastagnaIT commented on resources/lib/common/storage.py:
```diff
@@ -61,6 +61,14 @@ def commit(self):
+            import traceback
+            error(traceback.format_exc())
+
+
```
This is for debug purpose?
make a separate commit for this file

Smeulf commented on resources/lib/kodi/library.py:
```diff
@@ -373,29 +411,64 @@ def update_item(item_task, library_home):
-    Update the local Kodi library with new episodes of exported shows
+    if not _export_all_new_episodes_running():
+        common.log('Starting to export new episodes for all tv shows')
+        common.debug('library')
```
May be removed

Smeulf commented on resources/lib/kodi/library.py:
```diff
@@ -373,29 +411,64 @@ def update_item(item_task, library_home):
+    if not _export_all_new_episodes_running():
+        common.log('Starting to export new episodes for all tv shows')
+        common.debug('library')
+        common.debug(g.library())
```
May be removed

CastagnaIT commented on resources/lib/kodi/library.py:
```diff
@@ -322,6 +354,8 @@ def _add_to_library(videoid, export_filename):
             # it an appropriate videoid for later reference
             library_node[id_item] = {
                 'videoid': videoid.derive_parent(depth)}
+            if depth < 2:  # If Show or Season, save the nfo export status
```
I ask for info, instead of depth, it is not possible to verify the videoid value of the node?

Smeulf replied:
can be the mediatype of the parent videoid yes, and will avoid remember that for movies.

CastagnaIT commented on resources/lib/kodi/library.py:
```diff
@@ -363,7 +397,11 @@ def remove_item(item_task, library_home=None):
+        # Is there a way to simplify the lambda expression ?
+        common.remove_path(id_path, g.library(), lambda e: (e.keys() == ['videoid']
+                                                            or e.keys() == ['videoid', 'nfo_export']
+                                                            or e.keys() == ['nfo_export', 'videoid']))
```
you can try to replace the two line 403-404 with:
`or all(k in e.keys() for k in ['videoid', 'nfo_export'])`

Smeulf replied:
or len(set(e.keys()) - {'videoid', 'nfo_export'}) == 0 works!

CastagnaIT commented on resources/lib/kodi/library.py:
```diff
@@ -373,29 +411,64 @@ def update_item(item_task, library_home):
+            common.debug('mediatype {}'.format(library_item['videoid'].mediatype))
+            if library_item['videoid'].mediatype == common.VideoId.SHOW:
+                export_new_episodes(library_item['videoid'], False)
+            # add some randomness
```
non sense comment, use explicit comment info (same at commit description)
if it is used to not be banned from netflix server or anything else it should be explained

Smeulf replied:
Indeed

CastagnaIT commented on resources/lib/services/library_updater.py:
```diff
@@ -44,69 +59,61 @@ def is_idle(self):
-    def onScanStarted(self, library):
+        self.next_schedule = _compute_next_schedule()
+
+    def onScanStarted(self, lib):
```
the call interface use 'library' not 'lib'
why you have changed?

Smeulf replied:
Because library masks the libraty global name (import resources.lib.kodi.library as library).
So to avoid confusion.

CastagnaIT replied:
so is more correct to use the right interface property names
instead the "import" can change name as you wish like

from: import resources.lib.kodi.library as library
to: import resources.lib.kodi.library as lib_library

this variable is used only once if I've seen it right

CastagnaIT commented on resources/lib/services/library_updater.py:
```diff
@@ -44,69 +59,61 @@ def is_idle(self):
             self.startidle = 0
-        idletime = lastidle - self.startidle
-        return idletime >= 300
+        return (lastidle - self.startidle) >= 300
```
please leave included the variable 'idletime' as original

Smeulf replied:
Can be done :)

Smeulf commented:
About the locale_id, it whoud be called workraound more than fix, you're right.

The problem comes when the service changes a setting. Somehow, the locale_id value is dropped from the global storage, but I couldn't find why.

So as the values are not pure setting IMO, I moved them to the persistent storage. But it seems ther's some more case this error can appear, that's why I added extra debug.

May I propose 2 commits?:
- 1 for the full export new episode feature, including locale_id workaround
- 1 for extra locale_id debug

Will be easier for me to add the changes you asked :p

CastagnaIT commented:
well boy now understand better
you can make the latest changes then let's see

Smeulf commented:
Commits squashed, changes made as your review.

I also had to change some things in the cache: it's needed to refresh the metadata before we can analyse them. But because of the cache, they was never refreshed.

Plus the cache invalidate was triggering an error if not loaded before trying to invalidate it. So... Please take a look on those changes.

If you feel it must be done in a different way, please improve my code, as I didn't find any other way to proceed.

CastagnaIT commented:
avoid exceptions, instead handle the empty metadata returned so remove the local files

CastagnaIT commented:
I think it is more likely that the whole season is withdrawn or that the show no longer exists
so you should at least manage this cases:
-A single season no longer present (no season in metadata list)
-A show that no longer exists (metadata return empty)

Smeulf commented:
> -A single season no longer present (no season in metadata list)
If it's a single season show, and the season is removed, it should return an empty metadata as it's as if the show was removed. But I'll handle it because of the "should" ^^

> -A show that no longer exists (metadata return empty)
That will happend often.

Removing local files is not good IMO: people using trackt may get their episodes removed from tracking. Plus the function name is "export new episodes", not remove them ^^

I'd prefer to start with the new episodes only, get feedback from users, and if needed add a new option 'remove deleted episodes' later.

CastagnaIT commented:
okay, so for now you can handle them by skipping the show / season to get updates

Smeulf commented:
Hope it's good, I have no example to try :/

CastagnaIT commented on resources/lib/kodi/library.py:
```diff
@@ -452,19 +450,24 @@ def export_new_episodes(videoid, scan=True):
+            if scan:
+                common.debug('Notify service to update the library')
+                common.send_signal(common.Signals.LIBRARY_UPDATE_REQUESTED)
+    else:
```
wrong indent

Smeulf replied:
The else is for `if videoid.mediatype == common.VideoId.SHOW:` `else:`

If not a tvshow, debug message. I plan to add an 'export new episodes' context menu on tvshow, it's to ensure it won't be started for anything else (ie with a xbmc.executebuildin from ananother addon with wrong parameters)

Smeulf commented:
Added more logs for missing tv show or no season

CastagnaIT commented on resources/lib/cache.py:
```diff
@@ -337,7 +337,14 @@ def verify_ttl(self, bucket, identifier, cache_entry):
+                self.add(bucket, identifier, cache_entry['content'])
+            except:
+                pass
         # Remove from in-memory cache
```
can you show me the error log?

Smeulf replied:
If there's no disk cache, I have a CacheMiss() error. I rewritten that function, you should like it better :)

Smeulf replied:
`# To ensure removing disk cache, it must be loaded first or it will trigger an exception
        cache_filename = self._entry_filename(bucket, identifier)
        cache_exixts = os.path.exists(cache_filename)

        if on_disk and cache_exixts:
            cache_entry = self._get_from_disk(bucket, identifier)
            self.add(bucket, identifier, cache_entry['content'])

        # Remove from in-memory cache
        del self._get_bucket(bucket)[identifier]
        # Remove from disk cache if it exists

        if cache_exixts:
            os.remove(cache_filename)`

Smeulf commented:
I'm annoyed with the PR. Looks like the library.ndb2 was cleared :/ I have to make it work for some more time to ensure it's working.

Smeulf commented:
Plus I have a problem while exporting a single episode. That's making me crazy!

CastagnaIT commented:
cleared 😲😕there's no hurry
stop for a while, another day you will be able to see things better!

Smeulf commented:
Should be better! Finally!

Gonna test that for a week or so and report.

I hope there's no more library cleared issue...

Smeulf commented on resources/lib/kodi/library.py:
```diff
@@ -363,7 +421,15 @@ def remove_item(item_task, library_home=None):
+        # Reverted.
+        common.remove_path(id_path, g.library(), lambda e: (
+            e.keys() == ['videoid']
+            or len(set(e.keys()) - {'videoid', 'nfo_export'}) == 0
```
For compatibility in case some tried my branch

Smeulf commented:
@CastagnaIT sounds good to me, ready to merge.

Would you have a final review in case I missed something?

For me, it's ready to merge.

Smeulf commented:
Also, IMO it would be good to release with this commit before adding anything else (ie settings to db). Because as it changes the internal library db structure, reverting to the previous version will become pretty hard for people using the Kodi library.

Smeulf commented:
Closing pull request because of the current database work.
Will be reopened later with adapted code.

notoco commented:
What about upnext fix from this?

Smeulf commented:
Won't be forgotten for next release ;)
